src/report.rs

```diff
@@ -0,0 +1,256 @@
+//! Optionally writes an html-formatted summary report after running a load test.
+
+use crate::metrics;
+
+use std::collections::BTreeMap;
+use std::mem;
+
+use serde::Serialize;
+
+/// Defines the metrics reported about requests.
+#[derive(Debug, Clone, Serialize)]
+pub struct RequestMetric {
+    pub method: String,
+    pub name: String,
+    pub number_of_requests: usize,
+    pub number_of_failures: usize,
+    pub response_time_average: String,
+    pub response_time_minimum: usize,
+    pub response_time_maximum: usize,
+    pub requests_per_second: String,
+    pub failures_per_second: String,
+}
+
+/// Defines the metrics reported about responses.
+#[derive(Debug, Clone, Serialize)]
+pub struct ResponseMetric {
+    pub method: String,
+    pub name: String,
+    pub percentile_50: String,
+    pub percentile_60: String,
+    pub percentile_70: String,
+    pub percentile_80: String,
+    pub percentile_90: String,
+    pub percentile_95: String,
+    pub percentile_99: String,
+    pub percentile_100: String,
+}
+
+/// Helper to generate a single response metric.
+pub fn get_response_metric(
+    method: &str,
+    name: &str,
+    response_times: &BTreeMap<usize, usize>,
+    total_request_count: usize,
+    response_time_minimum: usize,
+    response_time_maximum: usize,
+) -> ResponseMetric {
+    // Calculate percentiles in a loop.
+    let mut percentiles = Vec::new();
+    for percent in &[0.5, 0.6, 0.7, 0.8, 0.9, 0.95, 0.99, 1.0] {
+        percentiles.push(metrics::calculate_response_time_percentile(
+            response_times,
+            total_request_count,
+            response_time_minimum,
+            response_time_maximum,
+            *percent,
+        ));
+    }
+
+    // Now take the Strings out of the Vector and build a ResponseMetric object.
+    ResponseMetric {
+        method: method.to_string(),
+        name: name.to_string(),
+        percentile_50: mem::take(&mut percentiles[0]),
+        percentile_60: mem::take(&mut percentiles[1]),
+        percentile_70: mem::take(&mut percentiles[2]),
+        percentile_80: mem::take(&mut percentiles[3]),
+        percentile_90: mem::take(&mut percentiles[4]),
+        percentile_95: mem::take(&mut percentiles[5]),
+        percentile_99: mem::take(&mut percentiles[6]),
+        percentile_100: mem::take(&mut percentiles[7]),
+    }
+}
+
+pub fn request_metrics_template(metric: RequestMetric) -> String {
```

Comment:
@LionsAd This is how we build the "inner" templates without using something like Handlebars to loop. format!() requires that the template itself has to be hard coded, we can't use a variable (or even a constant).